### ISSUE TITLE
Strict Concurrency: MainActor isolated KeyPath warning False Positive

### ISSUE THREAD

**Issue by tcldr** [apple/swift#72181]
### Description

When declaring a property wrapper on a MainActor isolated type that takes a MainActor isolated KeyPath, the 5.10 compiler flags a concurrency warning: 'Cannot form key path to main actor-isolated property 'requestReview'; this is an error in Swift 6'

### Reproduction

```swift
import StoreKit

@MainActor
struct MyView: View {
  // Cannot form key path to main actor-isolated property 'requestReview'; this is an error in Swift 6
  @Environment(\.requestReview) private var requestReview

  var body: some View {
    Text("Cannot form key path to main actor-isolated property")
  }
}
```

### Expected behavior

No warning.

### Environment

swift-driver version: 1.90.11.1 Apple Swift version 5.10 (swiftlang-5.10.0.13 clang-1500.3.9.4)
Target: arm64-apple-macosx14.0

### Additional information

First flagged here: https://forums.swift.org/t/swift-5-10-concurrency-and-xctest/69929/11

**Comment by wmorgue** [comment]
Xcode 16 still has this 🐞
swift-driver version: 1.109.2 Apple Swift version 6.0 (swiftlang-6.0.0.3.300 clang-1600.0.20.10)
Target: arm64-apple-macosx14.0

**Comment by NachoSoto** [comment]
Filed as https://github.com/swiftlang/swift/issues/74636 too

**Comment by NachoSoto** [comment]
Filed 4 months ago and still unfixed in Swift 6. Meaning you can't use this environment variable while compiling with Swift 6.

**Comment by radude89** [comment]
Are there any workarounds available?

**Comment by hborla** [comment]
Sorry I missed this - if that happens again, please feel free to add the `concurrency` label yourself, after which it'll get added to my queue of concurrency issues to montior/triage/fix.

**Comment by hborla** [comment]
@xedin I think isolated key paths need to understand isolated default values. This code is valid because `var requestReview`, and thus its initial value, are isolated to the main actor.

**Comment by hborla** [comment]
>Are there any workarounds available?

You should be able to work around this by initializing the property wrapper in a `@MainActor`-isolated initializer, like this

```swift
import SwiftUI
import StoreKit

@MainActor
struct MyView: View {
  @Environment private var requestReview: RequestReviewAction

  init() {
    self._requestReview = Environment(\.requestReview)
  }

  var body: some View { ... }
}
```

**Comment by NachoSoto** [comment]
@hborla as an external contributor I'm not allowed to add labels to issues I file unfortunately.

**Comment by hborla** [comment]
>@hborla I appreciate that, but as an external contributor I'm not allowed to add labels to issues I file unfortunately.

That is annoying! I didn't realize that. It's fine to tag me and I can triage the issue directly. I'm also looking to put a process in place to fill the gaps in issue triage, as I know this isn't the first time an issue has been missed.

**Comment by fritzt0** [comment]
Did this fix make it into Xcode 16 beta 4? I still have a very similar issue with a self-defined property wrapper and key path although the issue with @Environment(\.requestReview) is gone.

https://github.com/swiftlang/swift/issues/75616

**Comment by hborla** [comment]
>Did this fix make it into Xcode 16 beta 4?

No, this fix is not included in Xcode 16 Beta 4. It is also not included in Beta 5.

**Comment by MaximBazarov** [comment]
hey @hborla still have the same issue on `Version 16.1 beta (16B5001e)`

```swift 
@MainActor final class TestOnMainActor {
    var foo: Int
}

let x = \TestOnMainActor.foo

```

any idea when it's going to be included?

**Comment by hborla** [comment]
Xcode 16.1 Beta 1 has the same Swift compiler as Xcode 16 Beta 5 (`swiftlang-6.0.0.7.6` for both, which does not contain the fix).

**Comment by xedin** [comment]
@msdrigg it looks like you don't have `InferSendableFromCaptures` upcoming feature enabled and not running in `-swift-version 6` mode where it's enabled by default.

**Comment by MaximBazarov** [comment]
FYI. Fixed in Version 16.0 beta 6 (16A5230g)
